tls: throw an error on getLegacyCipher

jasnell opened the pull request:
throw an error on unknown getLegacyCipher input
rather than returning the default. Makes the
behavior more explicit and matches up with the
behavior on the command line switches.

jasnell commented:
@misterdjules @mdawsonibm

misterdjules commented:
Can we add a test to make sure that `getLecagyCiphers` throws/doesn't throw as expected?

jasnell commented:
Done

misterdjules commented:
@jasnell Do we want to throw an error on a call to `getLegacyCiphers()` when the first parameter passed is not a string (undefined, null or anything else that is not a string)? The error message would be different, instead of `'Error: Unknown legacy cipher list'`, it would be something like `'Error: getLegacyCiphers requires a string as first parameter"`.

Also, I think we miss the "full" documentation for the `getLegacyCiphers` function in the API docs, including what the parameter is, its type, etc. Currently it is only briefly mentioned in the `Modifying the Default Cipher Suite` section.

jasnell commented:
@misterdjules  ... PR updated to address feedback.

jasnell commented:
https://github.com/joyent/node/issues/15445 @misterdjules

misterdjules commented:
@jasnell Thank you, I think that's a significant improvement! :+1:

mdawsonibm commented:
One minor nit.  Should the tests validate the type of the Error, ie a TypeError or an Error as appropriate ? Otherwise lgtm.

jasnell commented:
/cc @mhdawson @misterdjules ok, please check the two additional commits. if these look good I'll squash the pr down and land.

misterdjules commented on test/simple/test-tls-cipher-list.js:
```diff
@@ -44,6 +44,21 @@ function doTest(checklist, env, useswitch) {
+      // Test NODE_LEGACY_CIPHER_LIST takes precendence over --cipher-list
+      args.unshift('--cipher-list=XYZ');
+      if (env) options = {env:{"NODE_LEGACY_CIPHER_LIST": env}};
+      break;
```
We would probably need to have more tests than that to test the whole matrix of command line switches and environment variables. I would do that in a separate PR though, as this one is about `ls: throw an error on getLegacyCipher`.

misterdjules commented:
@jasnell I think @mhdawson meant that we should check for the type of the error thrown by `tls.getLegacyCiphers`. I didn't see anything related to that in the latest two commits.

I would not add a lot to this PR so that we can reason about it easily and land something reasonably quickly. We can work on other aspects of ciphers list related changes in other PRs.

Thank you again very much for your work!

jasnell commented:
@misterdjules after talking with @mhdawson today, I realize now that I missed an important aspect of your discussion with him regarding the order of precedence on the command line switches and environment variables, as well as the error type check in test test case. I'm going to make those additional changes and squash these PR's down this evening with the goal of landing them by Friday.

jasnell committed:
tls: order of precedence on cipher list options, type checking

This commit squashes down several previous commits on this change
to clean up before landing. There are several changes to the
new cipher list command line switches:

1. It includes the type checking fix on getLegacyCiphers
2. It has the command line switches override the environment
   variables
3. It documents it order of precedence
4. Expands the test case to test for order of precedence

jasnell committed:
test: test-tls-cipher-list test error type

test that the type of error thrown is correct

jasnell commented:
@misterdjules @mhdawson ... ok the changes discussed have been made and the multiple commits have been squashed down into two distinct commits.

jasnell committed:
test: fix typo in tls.markdown doc and additional tests

Fix a minor typo in the tls.markdown and add additional
precedence tests in test-tls-cipher-list

jasnell committed:
tls: refactor test-tls-cipher-list and v0.10.38 legacy test

Per the feedback from Julien, refactor the test-tls-cipher-list
test to avoid using the confusing switch statement. Add tests
to ensure that using `--enable-legacy-cipher-list=v0.10.38`
or `NODE_LEGACY_CIPHER_LIST=v0.10.38` disables the use of the
default cipher list but setting the `--cipher-list` equal to
the v0.10.38 list explicitly does not.

Minor refactor to tls.js to ensure the above behavior.

misterdjules commented on lib/tls.js:
```diff
@@ -1338,7 +1352,8 @@ exports.connect = function(/* [port, host], options, cb */) {
     rejectUnauthorized: '0' !== process.env.NODE_TLS_REJECT_UNAUTHORIZED
   };
-  if (DEFAULT_CIPHERS != _crypto.getLegacyCiphers('v0.10.38')) {
+  if (!usingV1038Ciphers()) {
```
Maybe add a comment to clarify why there's an exception for v0.10.38 here? Otherwise I'm concerned it's going to be difficult to understand a few months/years from now.

misterdjules commented on doc/api/tls.markdown:
```diff
@@ -164,6 +164,35 @@ Reverting back to the defaults used by older releases can weaken the security
+
+Note that the `--enable-legacy-cipher-list`, `NODE_LEGACY_CIPHER_LIST`,
+`--cipher-list` and `NODE_CIPHER_LIST` options are mutually exclusive. Only
+_one_ should be used at a time.
```
I would suggest removing `Only _one_ should be used at a time`, since the support for having some options/env vars override others is intentional and corresponds to valid use cases.

misterdjules commented on test/simple/test-tls-cipher-list.js:
```diff
@@ -60,11 +45,185 @@ function doTest(checklist, env, useswitch) {
       });
 }
 
+// test that the command line switchs takes precedence
```
switchs -> switches.

misterdjules commented on test/simple/test-tls-cipher-list.js:
```diff
@@ -60,11 +45,185 @@ function doTest(checklist, env, useswitch) {
+  doTest(V1038Ciphers,
+         ['--enable-legacy-cipher-list=v0.10.38'],
+         {
+           'NODE_LEGACY_CIPHER_LIST': 'v0.10.38',
```
Don't we want another value for `NODE_LEGACY_CIPHER_LIST` (e.g `ABC`), otherwise how do we know that the value for legacy cipher list was taken from the command line switch and not from the environment variable?

jasnell replied:
In v0.10.39 there is only one legal value for this. Passing in anything but
v0.10.38 throws, even if it's overridden by the command line switch. For
v0.12.3 and master I specify different values.
On May 4, 2015 5:19 PM, "Julien Gilli" <notifications@github.com> wrote:

> In test/simple/test-tls-cipher-list.js
> <https://github.com/joyent/node/pull/14572#discussion_r29636669>:
>
> > +  // over NODE_CIPHER_LIST
> > +  doTest(V1038Ciphers,
> > +         ['--enable-legacy-cipher-list=v0.10.38'],
> > +         {'NODE_CIPHER_LIST': 'XYZ'});
> > +
> > +  // test that --cipher-list takes precedence over NODE_LEGACY_CIPHER_LIST
> > +  doTest('ABC',
> > +         ['--cipher-list=ABC'],
> > +         {'NODE_LEGACY_CIPHER_LIST': 'v0.10.38'});
> > +
> > +
> > +  // test that --enable-legacy-cipher-list takes precence over both envars
> > +  doTest(V1038Ciphers,
> > +         ['--enable-legacy-cipher-list=v0.10.38'],
> > +         {
> > +           'NODE_LEGACY_CIPHER_LIST': 'v0.10.38',
>
> Don't we want another value for NODE_LEGACY_CIPHER_LIST (e.g ABC),
> otherwise how do we know that the value for legacy cipher list was taken
> from the command line switch and not from the environment variable?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joyent/node/pull/14572/files#r29636669>.
>

misterdjules commented on test/simple/test-tls-cipher-list.js:
```diff
@@ -60,11 +45,185 @@ function doTest(checklist, env, useswitch) {
+    // test the right-most command line option takes precedence
+    doTest('XYZ',
+           [
+             '--cipher-list=XYZ',
```
Do we want another value for the first `--cipher-list` command line switch (e.g `ABC`)? Otherwise how do we know that the value for ciphers list was taken from the second `--cipher-list` and not the first one?

jasnell replied:
Yeah, actually that's just a typo... it should have been different

misterdjules commented on test/simple/test-tls-cipher-list.js:
```diff
@@ -60,11 +45,185 @@ function doTest(checklist, env, useswitch) {
+
+var test_count = 0;
+
+function doDefaultCipherTest(additional_args, env, failexpected) {
```
Minor: instead of passing booleans, I find passing options objects easier to read. In this specific case:
```
doDefaultCipherTest([], {}, test_uses_default_cipher_list);
```
turns into
```
doDefaultCipherTest([], {},  { failExpected: true });
```

misterdjules commented on test/simple/test-tls-cipher-list.js:
```diff
@@ -60,11 +45,185 @@ function doTest(checklist, env, useswitch) {
+  function() {
+    var tls = require('tls');
+    var orig_createCredentials = require('crypto').createCredentials;
+    require('crypto').createCredentials = function(options) {
```
Instead of monkey patching `crypto.createCredentials`, what do you think about starting a server that sets its ciphers list to `RC4-MD5` only?

Since it's explicitly excluded from the ciphers list after node v0.10.38, the test could make sure that the client connection can only be established when passing `--enable-legacy-cipher-list=v0.10.38` or setting equivalent environment variables.

I believe it would make the implementation of this test much simpler.

jasnell replied:
That's not quite what this is testing. This tests to make sure the default
ciphers are not set on the options when v0.10.38 legacy is set. The monkey
patch allows us to intercept the call and verify that the ciphers property
is indeed not set.
On May 4, 2015 6:05 PM, "Julien Gilli" <notifications@github.com> wrote:

> In test/simple/test-tls-cipher-list.js
> <https://github.com/joyent/node/pull/14572#discussion_r29638490>:
>
> > +
> > +// Test to ensure default ciphers are not set when v0.10.38 legacy cipher
> > +// switch is used. This is a bit involved... we need to first set up the
> > +// TLS server, then spawn a second node instance using the v0.10.38 cipher,
> > +// then connect and check to make sure the options are correct. Since there
> > +// is no direct way of testing it, an alternate createCredentials shim is
> > +// created that intercepts the call to createCredentials and checks the output.
> > +// The following server code was adopted from test-tls-connect-simple.
> > +
> > +// note that the following function is written out to a string and
> > +// passed in as an argument to a child node instance.
> > +var script = (
> > +  function() {
> > +    var tls = require('tls');
> > +    var orig_createCredentials = require('crypto').createCredentials;
> > +    require('crypto').createCredentials = function(options) {
>
> Instead of monkey patching crypto.createCredentials, what do you think
> about starting a server that sets its ciphers list to RC4-MD5 only?
>
> Since it's explicitly excluded from the ciphers list after node v0.10.38,
> the test could make sure that the client connection can only be established
> when passing --enable-legacy-cipher-list=v0.10.38 or setting equivalent
> environment variables.
>
> I believe it would make the implementation of this test much simpler.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joyent/node/pull/14572/files#r29638490>.
>

misterdjules replied:
OK, thank you for the clarification. I would suggest also asserting that the monkey-patched version of `crypto.createCredentials` was called to avoid missing potential future regressions if `crypto.createCredentials` ends up being refactored and is not called anymore (like in v0.12).

Also, do we really need to spawn a server to run that test? If I understand correctly, `crypto.createCredentials` is called before any connection is initiated, so I would expect that we can perform the same test without starting a server.

jasnell replied:
I spawn the server purposefully to do a full roundtrip to make sure that the connection still works without those defaults being set on the client side.

misterdjules replied:
@jasnell OK, I understand now. Generally, testing a single characteristic/behavior in a test helps make it clearer and avoid false negatives.

In this case, the test doesn't make a difference between the child process writing to stderr because the client could not connect or because the `options.ciphers` value passed to `crypto.createCredentials` was not expected. See my comment here: https://github.com/joyent/node/pull/14572/files#r30068923.

If I understand correctly, `doDefaultCipherTest` seems to test _more than one thing_. It tests that `crypto.createCredentials` is called with the proper value for its `options.ciphers` argument, _and_ that a client that uses `--enable-legacy-cipher-list` can connect to a `tls.Server` instance started with the default options.

I think splitting that in two different simpler tests would make the tests more reliable and easier to understand.

misterdjules commented:
@jasnell Added a few comments, but it's definitely getting better, thank you!

jasnell committed:
test: additional refinements in the test

Per the latest round of feedback from julien.
Fix a couple of typos, add some explanatory text,
refactor the test case.

misterdjules commented on test/simple/test-tls-cipher-list.js:
```diff
@@ -19,36 +19,21 @@
 // USE OR OTHER DEALINGS IN THE SOFTWARE.
 
-var spawn = require('child_process').spawn;
+var spawn  = require('child_process').spawn;
```
Style: the current style used does not align the `=` operator for assignment statements.

misterdjules commented on test/simple/test-tls-cipher-list.js:
```diff
@@ -60,11 +45,192 @@ function doTest(checklist, env, useswitch) {
+  doTest('ABC',
+         ['--cipher-list=ABC'],
+         {'NODE_LEGACY_CIPHER_LIST': 'v0.10.38'});
+
```
Style: unnecessary blank line.

misterdjules commented on test/simple/test-tls-cipher-list.js:
```diff
@@ -60,11 +45,192 @@ function doTest(checklist, env, useswitch) {
+
+    // test that NODE_LEGACY_CIPHER_LIST takes precedence over
+    // NODE_CIPHER_LIST
+
```
Style: unnecessary blank line.

jasnell committed:
test: additional round of edits on the test-tls-cipher-list.js

Per feedback from Julien.

jasnell commented:
@misterdjules @mhdawson

misterdjules commented:
@jasnell Thank you, I'll take a look on Sunday!

misterdjules commented on src/node_crypto.cc:
```diff
@@ -4199,12 +4199,18 @@ const char* ToCString(const node::Utf8Value& value) {
   HandleScope scope;
+  unsigned int len = args.Length();
+  if (len != 1 || !args[0]->IsString()) {
+    return ThrowException(Exception::TypeError(String::New("A single string parameter is required")));
```
Style: wrap at 80 columns.

misterdjules commented on test/simple/test-tls-cipher-list.js:
```diff
@@ -60,11 +45,194 @@ function doTest(checklist, env, useswitch) {
+      out += data;
+    }).
+    on('end', function() {
+      if (opts.failExpected && err === '') {
```
Why is this test done in the listener for the `'end'` event on the child process' standard output? The child seems to only write to the standard _error_ output.

misterdjules commented on test/simple/test-tls-cipher-list.js:
```diff
@@ -60,11 +45,194 @@ function doTest(checklist, env, useswitch) {
+      port: 0,
+      rejectUnauthorized: false
+    }, function() {
+      socket.end();
```
If the client cannot connect, an `'error'` event would be emitted and node would write to stderr. The parent process would check that some data was read from the child's stderr, and it would make the test pass, but it shouldn't.

misterdjules commented on test/simple/test-tls-cipher-list.js:
```diff
@@ -60,11 +45,194 @@ function doTest(checklist, env, useswitch) {
+  if (env) options.env = env;
+  var out = '', err = '';
+  additional_args = additional_args || [];
+  var args = additional_args.concat([
```
Another way to achieve the same thing would be to use `fork` and to pass an argument to the child process that allows to easily identify what code to run in the parent and in the child processes. Something similar to what is done in https://github.com/joyent/node/blob/v0.10/test/simple/test-domain-top-level-error-handler-throw.js. A search with `grep -rHn "'child'" test/simple` will show you many tests where this pattern is used.

That would require to have this test in a separate file, and I think this would make things simpler and cleaner overall.

jasnell commented:
@misterdjules ... I'll refactor that last test to split things out more. No problem there. Will see if I can get that in this morning. Beyond that, are there any other remaining issues on this?

jasnell committed:
test: refactor default cipher list test

Refactor the default cipher list test to separate
concerns a bit, make the test clearer based on
Julien's feedback.

Also, fix the 80-char wrap issue in the related
node_crypto.cc DefaultCiphers method.

jasnell commented:
@misterdjules

misterdjules commented:
@jasnell I would like to come up with a good solution for https://github.com/joyent/node/pull/23947 too, and add support for the new command line switches in `test/external/ssl-options`. I can add that since I'm familiar with how this test works.

We don't have to rush it though. As we discussed during the latest weekly meeting, if we want to release v0.12.3 before these changes are ready, we can revert the work in progress in v0.12 and include them back in time for the next v0.12.4 release.

jasnell commented:
Yes, I'm going to look at that one tonight again
On May 12, 2015 1:11 PM, "Julien Gilli" <notifications@github.com> wrote:

> @jasnell <https://github.com/jasnell> I would like to come up with a good
> solution for #23947 <https://github.com/joyent/node/pull/23947> too, and
> add support for the new command line switches in test/external/ssl-options.
> I can add that since I'm familiar with how this test works.
>
> We don't have to rush it though. As we discussed during the latest weekly
> meeting, if we want to release v0.12.3 before these changes are ready, we
> can revert the work in progress in v0.12 and include them back in time for
> the next v0.12.4 release.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joyent/node/pull/14572#issuecomment-101406567>.
>

misterdjules commented:
Moving to milestone v0.10.40, as v0.10.39 was released today.

misterdjules commented:
Closing in favor of #25603.
